pkg/process/maps.go

```diff
@@ -72,24 +91,51 @@ func (c *mappingFileCache) MappingForPID(pid int) ([]*profile.Mapping, error) {
 
 func (c *mappingFileCache) mappingForPID(pid int) ([]*profile.Mapping, error) {
 	mapsFile := fmt.Sprintf("/proc/%d/maps", pid)
-	h, err := hash.File(c.fs, mapsFile)
+
+	f, err := c.fs.Open(mapsFile)
 	if err != nil {
 		if os.IsNotExist(err) {
 			return nil, ErrNotFound
 		}
 		return nil, err
 	}
+	defer f.Close()
+
+	// About ~5% of CPU time of Parca Agent are spent in hashing object files
+	// to prepare for the debuginfo Exists gRPC call.
+	// This is even though files are extremely unlikely to change.
+	// We should only rehash if the mtime of the file changes.
+	var (
+		t int64
+		h uint64
+		r io.Reader
+	)
+	stat, err := f.Stat()
+	if err == nil && stat != nil {
+		t = stat.ModTime().UnixNano()
```

Comment:
do we have some docs or proof that `procfs` "files" (in particular mappings here of course) actually behave as expected in regards to mtime?

Reply:
this StackOverflow post suggests we cannot trust it: https://stackoverflow.com/a/30885400

Reply:
You're right! All the processes on my Linux machine but pid 1 show atime/mtime/ctime matching a time of the first `open` or `stat`.

```bash
$ stat /proc/2/maps
  File: /proc/2/maps
  Size: 0         	Blocks: 0          IO Block: 1024   regular empty file
Device: 16h/22d	Inode: 397849      Links: 1
Access: (0444/-r--r--r--)  Uid: (    0/    root)   Gid: (    0/    root)
Access: 2022-12-21 16:34:11.553962763 +0000
Modify: 2022-12-21 16:34:11.553962763 +0000
Change: 2022-12-21 16:34:11.553962763 +0000
 Birth: -

$ stat /proc/1/maps
  File: /proc/1/maps
  Size: 0         	Blocks: 0          IO Block: 1024   regular empty file
Device: 16h/22d	Inode: 25064       Links: 1
Access: (0444/-r--r--r--)  Uid: (    0/    root)   Gid: (    0/    root)
Access: 2022-11-29 21:26:10.799570796 +0000
Modify: 2022-11-29 21:26:10.799570796 +0000
Change: 2022-11-29 21:26:10.799570796 +0000
 Birth: -

$ sudo go run ./t/
2022-11-29 21:26:10.799570796 +0000 UTC
```

<details>

```go
package main

import (
	"fmt"
	"os"
)

func main() {
	f, err := os.Open("/proc/1/maps")
	if err != nil {
		panic(err)
	}
	s, err := f.Stat()
	if err != nil {
		panic(err)
	}
	fmt.Println(s.ModTime())
}
```

</details>

Reply:
> 你是对的！我的 Linux 机器上除 pid 1 之外的所有进程都显示 atime/mtime/ctime 匹配第一个时间`open`或`stat`.
> 
> ```shell
> $ stat /proc/2/maps
>   File: /proc/2/maps
>   Size: 0         	Blocks: 0          IO Block: 1024   regular empty file
> Device: 16h/22d	Inode: 397849      Links: 1
> Access: (0444/-r--r--r--)  Uid: (    0/    root)   Gid: (    0/    root)
> Access: 2022-12-21 16:34:11.553962763 +0000
> Modify: 2022-12-21 16:34:11.553962763 +0000
> Change: 2022-12-21 16:34:11.553962763 +0000
>  Birth: -
> 
> $ stat /proc/1/maps
>   File: /proc/1/maps
>   Size: 0         	Blocks: 0          IO Block: 1024   regular empty file
> Device: 16h/22d	Inode: 25064       Links: 1
> Access: (0444/-r--r--r--)  Uid: (    0/    root)   Gid: (    0/    root)
> Access: 2022-11-29 21:26:10.799570796 +0000
> Modify: 2022-11-29 21:26:10.799570796 +0000
> Change: 2022-11-29 21:26:10.799570796 +0000
>  Birth: -
> 
> $ sudo go run ./t/
> 2022-11-29 21:26:10.799570796 +0000 UTC
> ```
> 
> ```go
> package main
> 
> import (
> 	"fmt"
> 	"os"
> )
> 
> func main() {
> 	f, err := os.Open("/proc/1/maps")
> 	if err != nil {
> 		panic(err)
> 	}
> 	s, err := f.Stat()
> 	if err != nil {
> 		panic(err)
> 	}
> 	fmt.Println(s.ModTime())
> }
> ```

i think the /proc/pid/maps file's mod time is expected to equal with access time.

and it is not allowed to modify:
```
sudo chmod 0755 /proc/19177/maps
chmod: changing permissions of ‘/proc/19177/maps’: Operation not permitted
```

i have a log file, the stat output meets i expect.
```
$ stat otel-collector.log
  File: ‘otel-collector.log’
  Size: 9145273667      Blocks: 17861872   IO Block: 4096   regular file
Device: fd11h/64785d    Inode: 1835016     Links: 1
Access: (0600/-rw-------)  Uid: (    0/    root)   Gid: (    0/    root)
Access: 2022-12-20 17:24:30.243937079 +0800
Modify: 2022-12-22 12:10:03.450355314 +0800
Change: 2022-12-22 12:10:03.450355314 +0800
 Birth: -
$ stat otel-collector.log
  File: ‘otel-collector.log’
  Size: 9145282651      Blocks: 17861896   IO Block: 4096   regular file
Device: fd11h/64785d    Inode: 1835016     Links: 1
Access: (0600/-rw-------)  Uid: (    0/    root)   Gid: (    0/    root)
Access: 2022-12-20 17:24:30.243937079 +0800
Modify: 2022-12-22 12:10:05.482427849 +0800
Change: 2022-12-22 12:10:05.482427849 +0800
 Birth: -
$ pwd
/proc/19177/root/home/data/logs/takumi
```